### PR TITLE
Parallel Jetifier

### PR DESCRIPTION
**Pull request by m4tt72**
Jetify files in parallel using node `child_process`.
Removing the `console.log` from the loop makes the `jetifier` faster

### CODE REVIEW

File: src/worker.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      if (data.includes(mode === 'forward' ? oldClass : newClass)) {
+        data = mode === 'forward' ?
+          data.replace(new RegExp(oldClass, 'g'), newClass) :
+          data.replace(new RegExp(newClass, 'g'), oldClass);
+        writeFileSync(file, data, { encoding: 'utf8' });
+      }
+    }
+  }
+
+  console.log(`${mode}-jetifyied ${filesChunk.length} file(s)`);
```
Comment (mikehardy) [inline src/worker.js]: should just be jetified I think? Because…english 😆 (rule, who needs rules, we like exceptions)

**Comment by mikehardy** [comment]
This seems cool but CI is choking on the ‘node’ version (that is ‘current’, which is v12.5 I think right now) doing the JS bundle. That seems like it shouldn’t have anything to do with this(!?), so I’ll investigate that

I will also see if I can add some sort of performance test (likely just wrapping the jetify call in `time`) so we can see what sort of boost this offers. I do like performance boosts but want to make sure the gain matches complexity in general - stability and longevity (which is related to a module’s ability to attract collaborators that can contribute easily) are goals of mine. I think this PR still looks reasonable from that perspective just want to make sure it really is achieving the gain that going parallel should achieve

File: src/worker.js (hunk 1 above)
Comment (m4tt72) [reply src/worker.js]: Haha yeah my mistake! it should be `jetified`... We are adding to both React Native and the dictionary

**Comment by m4tt72** [comment]
Yes a performance test would be helpful! 
On my machine, I gained +3 seconds when running the jetifier in parallel. In our project it went from 4 seconds to 800ms! not a huge performance boost but I like to squeeze every ms out of it...

**Comment by mikehardy** [comment]
>3000 to 800ms is at minimum a 4x speedup, that’s motivating! I’m this close to smacking the approve button I just want to get CI working :-). Ah, CI problems.

**Comment by m4tt72** [comment]
@mikehardy the build fails probably because in the install script you are using `npm` instead of `yarn`. Try switching to `yarn` and see if that fixes the issue

**Comment by mikehardy** [comment]
Nah, relying on a package-manager difference to make things work is voodoo I think. npm and yarn are both capable (i use both, depending on the project, but npm in my biggest one - it’s fine). 

it’s failing to bundle in RN0.60, not failing to jetify, something else is going on. I haven’t investigated yet, it’s still on the list as I work through items today. It will probably be simple, and I need to update rn-androidx-demo to RN0.60-RC3 anyway

File: index.js [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
-    }
-  }
-}
\ No newline at end of file
+console.log(`Found ${files.length} file(s)...`);
+
+for (const filesChunk of chunk(files, cpus)) {
+  const worker = fork(join(__dirname, 'src', 'worker.js'));
+  worker.send({ filesChunk, classesMapping, mode });
+}
```
[the remark this answers is not in the document]
Comment (m4tt72) [reply index.js]: Adding a log right after the loop that spawns the `workers` will immediately execute. Best solution is to leave the logs inside the worker

[the remark this answers is not in the document]
Comment (mikehardy) [reply index.js]: Oh of course! spawn goes async by default. I still don’t like the O(cpus) pile of log statements - I don’t think they particularly add much, and as long as you know jetifier triggered (which the first statement will do) you’ll know that’s what is working until the next thing happens.So I’d learn towards removing the worker logs totally and just going with the initial log unless it was easy to join on all the workers exit events and print a final statement (note: I read up on the docs, id didn’t look worth it to me)

[the remark this answers is not in the document]
Comment (mikehardy) [reply index.js]: BTW - I reliably reproduce your 4x speedup, and I’m nearly done exposing the node-pre-gyp fault on mac in CI plus the RN0.60 thing. So it’s all moving forward anyway.

[the remark this answers is not in the document]
Comment (m4tt72) [reply index.js]: Yep, that's what I did, removed the worker logs completely and left the initial log!

**Comment by mikehardy** [comment]
Just for reference for later - the current rn-androidx-demo test suite is running at 3.4-4s forward and (because the build folder is huge) about 20s in reverse

Once I commit the CI items and merge/re-run this we'll have the competitive timings from the update. I predict ~1s forward and 5s reverse.

**Comment by mikehardy** [comment]
Well, timing was right on for scaling by CPUs with 2 CPUs, so in CI it was 2x faster - ~2s and ~10s. Nice to see it scales like that actually. This is a great addition, I can’t think of a single reason at this point why someone would object to relying on this to ease the AndroidX pain. Thanks @m4tt72

**Pull request merged**